src/OpenTelemetry/Logs/LogRecordPool.cs

```diff
@@ -0,0 +1,178 @@
+// <copyright file="LogRecordPool.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.Threading;
+using OpenTelemetry.Internal;
+
+namespace OpenTelemetry.Logs
+{
+    /// <summary>
+    /// Manages a pool of <see cref="LogRecord"/> instances.
+    /// </summary>
+    public sealed class LogRecordPool
+    {
+        private const int DefaultMaxPoolSize = 1024;
+
+        private static LogRecordPool current = new(DefaultMaxPoolSize);
+
+        [ThreadStatic]
+        private static LogRecord? threadStaticLogRecord;
+
+        private readonly int sharedPoolSize;
+        private readonly LogRecord?[] sharedPool;
+        private int sharedPoolCurrentIndex = -1;
+
+        private LogRecordPool(int size)
+        {
+            this.sharedPoolSize = size;
+            this.sharedPool = new LogRecord?[size];
+        }
+
+        /// <summary>
+        /// Resize the pool.
+        /// </summary>
+        /// <param name="size">The maximum number of <see cref="LogRecord"/>s to store in the pool.</param>
+        public static void Resize(int size)
+        {
+            Guard.ThrowIfOutOfRange(size, min: 1);
+
+            current = new LogRecordPool(size);
+        }
+
+        /// <summary>
+        /// Rent a <see cref="LogRecord"/> from the pool.
+        /// </summary>
+        /// <returns><see cref="LogRecord"/>.</returns>
+        public static LogRecord Rent() => current.RentCore(clearIfReused: true);
```

Comment:
I don't have a clear position right now - would we prefer to do the clean up here or earlier (e.g. when the ref count goes to zero)? There might be small perf differences (e.g. I imagine `Rent()` is always called synchronously, while DecRef might be called asynchronously - from the exporter thread).

Reply:
In the new iteration there isn't much to clear, log data is set fully each time (overriding what was there before). There is a bit of cleanup for the new reusable tag/attribute storage which I put on the return to the pool.